tests/Helpers/JWKFetcherTest.php

```diff
@@ -0,0 +1,195 @@
+    {
+        $jwksFetcher = $this->getStub();
+
+        $pem = $jwksFetcher->getJwksX5c( 'https://localhost/.well-known/jwks.json', '__test_kid_2__' );
+
+        $this->assertNotEmpty($pem);
+
+        $pem_parts = explode( PHP_EOL, $pem );
+
+        $this->assertEquals( 4, count($pem_parts) );
+        $this->assertEquals( '-----BEGIN CERTIFICATE-----', $pem_parts[0] );
+        $this->assertEquals( '-----END CERTIFICATE-----', $pem_parts[2] );
+        $this->assertEquals( '__test_x5c_2__', $pem_parts[1] );
+    }
+
+    /**
+     * Test that a custom JWKS path returns the correct JSON.
+     *
+     * @return void
+     */
+    public function testGetJwksX5cWithPath()
+    {
+        $jwksFetcher = $this->getStub();
+
+        $pem = $jwksFetcher->getJwksX5c( 'https://localhost/.custom/jwks.json', '__test_custom_kid__' );
+
+        $this->assertNotEmpty($pem);
+
+        $pem_parts = explode( PHP_EOL, $pem );
+
+        $this->assertEquals( 4, count( $pem_parts ) );
+        $this->assertEquals( '-----BEGIN CERTIFICATE-----', $pem_parts[0] );
+        $this->assertEquals( '-----END CERTIFICATE-----', $pem_parts[2] );
+        $this->assertEquals( '__test_custom_x5c__', $pem_parts[1] );
+    }
+
+    /**
+     * Test that a JWK with no x509 cert returns a blank array.
+     *
+     * @return void
+     */
+    public function testGetJwksX5cNoX5c()
+    {
+        $jwksFetcher = $this->getStub();
+
+        $pem = $jwksFetcher->getJwksX5c( 'https://localhost/.custom/jwks.json', '__no_x5c_test_kid_2__' );
+
+        $this->assertEmpty($pem);
+    }
+
+    /**
+     * Test that the protected getProp method returns correctly.
+     *
+     * @return void
+     */
+    public function testConvertCertToPem()
+    {
+        $class  = new \ReflectionClass(JWKFetcher::class);
+        $method = $class->getMethod('convertCertToPem');
+        $method->setAccessible(true);
+
+        $test_string_1 = '';
+        for ($i = 1; $i <= 64; $i++) {
+            $test_string_1 .= 'a';
+        }
+
+        $test_string_2 = '';
+        for ($i = 1; $i <= 64; $i++) {
+            $test_string_2 .= 'b';
+        }
+
+        $returned_pem = $method->invoke(new JWKFetcher(), $test_string_1.$test_string_2);
+        $pem_parts    = explode( PHP_EOL, $returned_pem );
+        $this->assertEquals( 5, count($pem_parts) );
+        $this->assertEquals( '-----BEGIN CERTIFICATE-----', $pem_parts[0] );
+        $this->assertEquals( $test_string_1, $pem_parts[1] );
+        $this->assertEquals( $test_string_2, $pem_parts[2] );
+        $this->assertEquals( '-----END CERTIFICATE-----', $pem_parts[3] );
+    }
+
+    public function testCacheReturn()
+    {
+        $cache_handler = new TestCacheHandler();
+        $jwksFetcher   = $this->getStub($cache_handler);
+
+        $jwks_url = 'https://localhost/.well-known/jwks.json';
+        $kid      = '__test_kid_2__';
+
+        $pem = $jwksFetcher->getJwksX5c( $jwks_url, $kid );
```

Comment:
before this line I'd make sure that cache for that key is not defined so test is fully independent: 

        `$cache_handler->set( $jwks_url.'|'.$kid, null );`